### ISSUE TITLE
Alternative Recovery Image

### ISSUE THREAD

**Issue by remay** [remay/tmf-hudl#3]
Would finding a CWM or TWRP recovery be useful?  Certainly some better backup capability would be extremely useful.

**Comment by AndrewChqng** [comment]
Found this which is a CWM that works pretty well
https://www.modaco.com/topic/366202-clockworkmod-recovery-v6031-for-rkflashtool-built-by-hal2710

**Comment by remay** [comment]
Thanks @Tech0308.  When I get a chance I'll see how that version compares with the version I found last year

**Comment by remay** [comment]
Note: When installing a new recovery image the  `/etc/install-recovery.sh` needs modifying/removing so that it doesn't re-write the recovery partition back to the original recovery image.

**Comment by RayyanAnsari** [comment]
however, internal storage on this custom recovery doesn't work.

**Comment by AndrewChqng** [comment]
Oh yeah, I don't really notice it because I just adb sideload everything😅.
I think it's because it expects a second partition for the internal sdcard so the files are there at /data/media/0...

**Comment by remay** [comment]
If that's all it is then could probably fix that with a soft-link, or a change to fstab?

**Comment by RayyanAnsari** [comment]
I'm working on a port of TWRP, and I have internal storage working.
Only thing that I need to fix is touch - it displays on the screen fine, however I have to use a USB mouse.

I'm building from source, so this is the latest TWRP v3.5.2.

**Comment by remay** [comment]
Assume you've seen this old thread: https://forum.freaktab.com/forum/development-area/rom-hacks-and-mods-development/10905-unofficial-twrp-2-7-0-0-for-rockchip-s/page38

and especially messages numbers 583 & 609 - but I can't find a conclusion to the discussion.

**Comment by RayyanAnsari** [comment]
Yep, what's annoying on FreakTab is that search doesn't work properly.
However, this CaptainMidnight person is the same person that made the CWM recovery linked above.
The CWM recovery does respond to touch, so I'll have to unpack that and see what they did.

**Comment by remay** [comment]
Google search with "site:forum.freaktab.com" as the first search term works quite well.

**Comment by RayyanAnsari** [comment]
@CaptainMidnight seems to be active on GitHub.

It's the same profile picture so it might be worth asking.

**Comment by CaptainMidnight** [comment]
Hi @WaluigiWare64 @remay what exactly are you looking for, both hudl and hudl2 were tablets I used to have. Only the hudl was capable of having a custom ROM as the hudl2 has a locked protected boot system.

I may still have both a custom based on v1.3, I think, as well as maybe some twrp etc stuff from back then on a backup NAS

**Comment by RayyanAnsari** [comment]
Hi @CaptainMidnight, we're interested in the original Hudl.
Browsing old forum posts on FreakTab, it appears you created a port of CWM for the Hudl.
I'm currently porting the latest version of TWRP to the Hudl and it has internal storage working, however, touch doesn't work.

Your CWM port seems to respond to touch. Would you happen to remember what you had to do to get it working?
> I may still have both a custom based on v1.3, I think, as well as maybe some twrp etc stuff from back then on a backup NAS

If you have a custom ROM that has a newer Android version than stock (4.2.2), and if you have the sources/image of a TWRP or CWM that has touch working then I would be interested!


re: touch not working with my TWRP build.
I don't really know what the issue is here, I know the touch device is `/dev/input/event2` because I ran `cat /dev/input/event2`then touched the touchscreen and saw output from the terminal.

**Comment by RayyanAnsari** [comment]
Actually, I just tried swiping around and it selected something. I'll have to look into this further.

EDIT: the hashes of the kernel in recovery and kernel in boot are different. Perhaps the touch events from the recovery kernel may be broken. I'm currently trying with the kernel from boot.

**Comment by CaptainMidnight** [comment]
Possibly I may have both TWRP and CWM, but definitely should have the latter.

I didn't produce those recovery builds, just helped with some testing. Another user 'blinkydamo' if I remember rightly created his v1.3 ROM which also includes the CWM recovery.

Personally at that time, I'd setup my own custom kitchen and created a couple of test builds but then hudl2 was launched etc etc.

I'll have to power up my archive NAS server at the weekend and have a look at what what's there still, I do think the hudl stuff is still there but will check.

In the meantime, if you look on modaco in the original hudl forum or freaktab you should find it as well.

[Helios Rom v1](https://www.modaco.com/topic/366953-helios-rom-v10/)

**Comment by CaptainMidnight** [comment]
I've had a look at what is on the NAS, it's the CWM recovery which is exactly the same as the one used in the Helios ROM.

From memory at the time I was going to rebuild the Helios ROM but base it on ver 1.3 of the stock hudl ROM - I can't remember what version of android that uses, but it shouldn't be too difficult to rebuild that along the lines of Helios ROM. You'd just need to reconfigure the image partitions as I know the stock hudl ROMs have recovery.img size issues....... either way the Helios and a stock v1.3v ROM with the addition of the freak tab kitchen would have everything you'd need.

I never fully got the TWRP recovery working and then the hudl2 came out, so just got left.

**Comment by remay** [comment]
If anyone wants the original Helios ROM then I have it.   I took some of the bits from that ROM and re-integrated them with the last stock (V1.3) ROM, and that's what makes up the current release from this repo.  I didn't re-integrate CWM recovery or anything else that significantly deviated from the Stock ROM, as the intention at that time was to make something very close to Stock.   @CaptainMidnight  thank you for all the work you and your predecessors did back then - most of what I did was just re-hashing you old work.

**Comment by RayyanAnsari** [comment]
Thanks to the helpful advice of the people on the TWRP chat, I have gotten the touchscreen working!
I'll release this once I iron out the internal storage support and add extra partitions to the fstab file.

**Comment by RayyanAnsari** [comment]
This is a debug version of my TWRP port. It's WIP so be careful if you have important data, but I have tested it myself.
Here is the checklist for inclusion at TWRP. All the 'Blocking checks' are fine, but if any of you have spare SD cards, I would be grateful if you could check these.
If there are any bugs you find, please let me know!


Blocking checks
- [x] Correct screen/recovery size
- [x] Working Touch, screen
- [x] Backup to internal/microSD (don't have SD card to test, internal is working)
- [x] Restore from internal/microSD (don't have SD card to test, internal is working)
- [x] reboot to system
- [x] ADB

Medium checks
- [ ] update.zip sideload (i have no update.zip to test with)
- [x] UI colors (red/blue inversions)
- [x] Screen goes off and on
- [x] F2FS/EXT4 Support, exFAT/NTFS where supported (F2FS, exFAT and NTFS not supported by kernel)
- [x] all important partitions listed in mount/backup lists
- [x] backup/restore to/from external (USB-OTG) storage
- [x] backup/restore to/from adb (https://gerrit.omnirom.org/#/c/15943/)
~~- [ ] decrypt /data~~ (/data is not encrypted on my device and there is no option in the Security settings)
- [x] Correct date

Minor checks
- [x] MTP export
- [x] reboot to bootloader
- [x] reboot to recovery
- [x] poweroff
- [x] battery level
~~- [ ] temperature~~ (NOT SUPPORTED)
- [x] encrypted backup
- [x] input devices via USB (USB-OTG) - keyboard, mouse
- [x] USB mass storage export
- [x] set brightness
~~- [ ] vibrate~~ (NOT SUPPORTED)
- [ ] screenshot (not working, causes crash)
- [ ] partition SD card (no sd card to test with)

[recovery-twrp.zip](https://github.com/remay/tmf-hudl/files/7296991/recovery-twrp.zip)

**Comment by remay** [comment]
@RayyanAnsari I think you wanted the default parameter file?  Is this what you're after?
````
FIRMWARE_VER:4.1.1
MACHINE_MODEL:rk30sdk
MACHINE_ID:007
MANUFACTURER:RK30SDK
MAGIC: 0x5041524B
ATAG: 0x60000800
MACHINE: 3066
CHECK_MASK: 0x80
PWR_HLD: 0,0,A,0,1
KERNEL_IMG: 0x60408000
#RECOVER_KEY: 1,1,0,20,0
CMDLINE:console=ttyFIQ0 androidboot.console=ttyFIQ0 board.ap_has_alsa=0 init=/init initrd=0x62000000,0x00800000 mtdparts=rk29xxnand:0x00002000@0x00002000(misc),0x00004000@0x00004000(kernel),0x00008000@0x00008000(boot),0x00010000@0x00010000(recovery),0x00020000@0x00020000(backup),0x00100000@0x00040000(cache),0x00002000@0x00140000(kpanic),0x00200000@0x00142000(system),0x00020000@0x00342000(factory),-@0x00362000(userdata)

**Comment by RayyanAnsari** [comment]
I found that you already posted it somewhere else.

**Comment by Normoa** [comment]
Does anyone have the necessary to root a Hudl2 (that I've been given)?

I've tried other links, but they are all dead.

**Comment by remay** [comment]
> Does anyone have the necessary to root a Hudl2 (that I've been given)?
> 
> I've tried other links, but they are all dead.

I am not aware of anyone ever rooting a Hudl2.

**Comment by Normoa** [comment]
I think you may be thinking of after the update to Lollipop.

This is a stock Kitkat Hudl2.
